src/libmongoc/src/mongoc/mongoc-shared.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2021 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "./mongoc-shared-private.h"
+
+#include "common-thread-private.h"
+#include <bson/bson.h>
+
+#include <assert.h>
+
+typedef struct _mongoc_shared_ptr_aux {
+   int refcount;
+   void (*dtor) (void *);
+   void *managed;
+} _mongoc_shared_ptr_aux;
+
+static void
+_release_aux (_mongoc_shared_ptr_aux *aux)
+{
+   aux->dtor (aux->managed);
+   bson_free (aux);
+}
+
+static bson_mutex_t g_shared_ptr_mtx;
+static bson_once_t g_shared_ptr_mtx_init_once = BSON_ONCE_INIT;
+
+static BSON_ONCE_FUN (_init_mtx)
+{
+   bson_mutex_init (&g_shared_ptr_mtx);
+}
+
+static void
+_shared_ptr_spin_lock ()
+{
+   bson_mutex_lock (&g_shared_ptr_mtx);
+}
+
+static void
+_shared_ptr_spin_unlock ()
+{
+   bson_mutex_unlock (&g_shared_ptr_mtx);
+}
+
+void
+mongoc_shared_ptr_rebind_raw (mongoc_shared_ptr *const ptr,
+                              void *const pointee,
+                              void (*const dtor) (void *))
+{
+   assert (ptr && "NULL given to mongoc_shared_ptr_rebind_raw()");
```

Comment:
```suggestion
   BSON_ASSERT_PARAM (ptr);
```
Should `assert()` be replaced with `BSON_ASSERT` and variants? Question also applies to all other use of `assert()` in this PR.